cmd/korg/audit.go

```
@@ -0,0 +1,336 @@
+}
+
+type Values struct {
+	Items [][]interface{} `json:"values,omitempty"`
+}
+
+type Frames struct {
+	Schema map[string]interface{} `json:"schema,omitempty"`
+	Data   Values                 `json:"data,omitempty"`
+}
+
+type DevStatsRequest struct {
+	Queries []Query `json:"queries"`
+}
+
+type Query struct {
+	RefID        string `json:"refId"`
+	DatasourceID int    `json:"datasourceId"`
+	RawSQL       string `json:"rawSql"`
+	Format       string `json:"format"`
+}
+
+type AuditOptions struct {
+	Period            string
+	ActivityThreshold int
+	Output            string
+	OutputFile        string
+	ExceptionsFile    string
+	CheckOwners       bool
+}
+
+type UserInfo struct {
+	Username      string
+	Contributions int
+	Orgs          []string
+	IsOwner       bool
+}
+
+type Exception struct {
+	Username string
+	Reason   string
+}
+
+func GetAllUsersInOrgs(o Options, orgs []string) (map[string][]string, error) {
+	userOrgs := map[string][]string{}
+	for _, org := range orgs {
+		relativeConfigPath := fmt.Sprintf("%s/config/%s/org.yaml", o.RepoRoot, org)
+
+		config, err := readConfig(relativeConfigPath)
+		if err != nil {
+			return userOrgs, err
+		}
+
+		for _, user := range append(config.Members, config.Admins...) {
+			userOrgs[user] = append(userOrgs[user], org)
+		}
+	}
+
+	return userOrgs, nil
+}
+
+func GetContributionsForAYear(period string) (map[string]Contribution, error) {
+	postBody := DevStatsRequest{
+		Queries: []Query{
+			{
+				RefID:        "A",
+				DatasourceID: 1,
+				RawSQL: fmt.Sprintf(`select
+  sub."Rank",
+  sub.name as name,
+  sub.value
+from (
+  select row_number() over (order by sum(value) desc) as "Rank",
+    split_part(name, '$$$', 1) as name,
+    sum(value) as value
+  from
+    shdev
+  where
+    series = 'hdev_contributionsallall'
+    and period = '%s'
+  group by
+    split_part(name, '$$$', 1)
+) sub`, period),
+				Format: "table",
+			},
+		},
+	}
+
+	requestBody, _ := json.Marshal(postBody)
```

Is there a reason we are ignoring the error here?